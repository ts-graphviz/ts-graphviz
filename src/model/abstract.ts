/**
 * Classes implemented in the 'ts-graphviz' library are designed to inherit from this class.
 * @category Domain Model
 */
export abstract class GraphvizObject {}

/**
 * Classes implemented in the 'ts-graphviz' library that implement the `toDot` method are designed to inherit from this class.
 * @category Domain Model
 */
export abstract class DotObject extends GraphvizObject {}
