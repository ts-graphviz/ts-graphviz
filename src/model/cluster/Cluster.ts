import { GraphvizObject, IDot } from '../../common';
import { SubgraphAttributes } from '../attributes';
import { Attributes } from '../attributes/Attributes';
import { EdgeAttributes } from '../attributes/EdgeAttributes';
import { NodeAttributes } from '../attributes/NodeAttributes';
import { Edge } from '../Edge';
import { Node } from '../Node';

// tslint:disable: max-classes-per-file

export type GraphType = 'digraph' | 'graph' | 'subgraph';

/**
 * @hidden
 */
export abstract class Cluster<ATTR extends Attributes> extends GraphvizObject implements IDot {
  public abstract readonly type: GraphType;
  public readonly graph: ATTR;
  public readonly edge = new EdgeAttributes();
  public readonly node = new NodeAttributes();

  private nodes: Map<string, Node> = new Map();

  private edges: Set<Edge> = new Set();

  private subgraphs: Map<string, Subgraph> = new Map();

  constructor(public id: string, attributes: ATTR) {
    super();
    this.graph = attributes;
  }

  public abstract toDot(): string;

  public createSubgraph(id: string): Subgraph {
    const graph = new Subgraph(id);
    this.subgraphs.set(id, graph);
    return graph;
  }

  public createNode(id: string, attributes?: NodeAttributes): Node {
    const node = new Node(id, attributes);
    this.nodes.set(id, node);
    return node;
  }

  public createEdge(from: Node, to: Node, attributes: EdgeAttributes): Edge {
    const edge = new Edge(from, to, attributes);
    this.edges.add(edge);
    return edge;
  }
}
/**
 * @category Primary
 */
export class Subgraph extends Cluster<SubgraphAttributes> {
  public type: GraphType = 'subgraph';

  constructor(id: string, attributes: SubgraphAttributes = new SubgraphAttributes()) {
    super(id, attributes);
  }

  public isSubgraphCluster(): boolean {
    return this.id.startsWith('cluster_');
  }

  public toDot(): string {
    throw new Error('Method not implemented.');
  }
}
