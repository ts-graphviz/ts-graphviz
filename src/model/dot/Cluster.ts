import { Attributes } from '../attributes/Attributes';
import { EdgeAttributes } from '../attributes/EdgeAttributes';
import { NodeAttributes } from '../attributes/NodeAttributes';
import { Edge } from '../Edge';
import { IDot } from '../interface';
import { Node } from '../Node';

import { SubgraphAttributes } from '../attributes';

// tslint:disable: max-classes-per-file

export type GraphType = 'digraph' | 'graph' | 'subgraph';

export abstract class Cluster<ATTR extends Attributes> implements IDot {
  public abstract readonly type: GraphType;
  public readonly graph: ATTR;
  public readonly edge = new EdgeAttributes();
  public readonly node = new NodeAttributes();

  private nodes: Map<string, Node> = new Map();

  private edges: Set<Edge> = new Set();

  private subgraphs: Map<string, Subgraph> = new Map();

  constructor(public id: string, attributes: ATTR) {
    this.graph = attributes;
  }

  public abstract toDot(): string;

  public createSubgraph(id: string): Subgraph {
    const graph = new Subgraph(id);
    this.subgraphs.set(id, graph);
    return graph;
  }

  public createNode(id: string, attributes?: NodeAttributes): Node {
    const node = new Node(id, attributes);
    this.nodes.set(id, node);
    return node;
  }

  public createEdge(from: Node, to: Node, attributes: EdgeAttributes): Edge {
    const edge = new Edge(from, to, attributes);
    this.edges.add(edge);
    return edge;
  }
}

export class Subgraph extends Cluster<SubgraphAttributes> {
  public type: GraphType = 'subgraph';

  constructor(id: string, attributes: SubgraphAttributes = new SubgraphAttributes()) {
    super(id, attributes);
  }

  public isSubgraphCluster(): boolean {
    return this.id.indexOf('cluster_') === 0;
  }

  public toDot(): string {
    throw new Error('Method not implemented.');
  }
}
