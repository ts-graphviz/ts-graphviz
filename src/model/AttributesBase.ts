import { DotBase } from '../abstract';
import { IAttributesBase } from '../types';
import { ID } from './ID';
/**
 * @hidden
 */
export abstract class AttributesBase<T extends string> extends DotBase implements IAttributesBase<T> {
  /** @hidden */
  protected attrs: Map<T, ID> = new Map();
  /** The size of the attribute. */
  get size(): number {
    return this.attrs.size;
  }
  /** The size of the attribute. */
  public get(key: T): ID | undefined {
    return this.attrs.get(key);
  }
  /** Set a value to the attribute. */
  public set(key: T, value: string | boolean | number | ID): void {
    if (value instanceof ID) {
      this.attrs.set(key, value);
    } else {
      this.attrs.set(key, new ID(value));
    }
  }

  public apply(attributes: { [key in T]?: string | boolean | number | ID }) {
    for (const [key, value] of Object.entries(attributes)) {
      this.set(key as T, value as string | boolean | number | ID);
    }
  }
}
