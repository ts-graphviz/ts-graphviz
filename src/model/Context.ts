import { Digraph, Graph, RootClusterType } from './cluster';

export interface IContext {
  graphType: RootClusterType;
}

export class Context implements IContext {
  get graphType(): RootClusterType {
    return this.root.type;
  }
  constructor(public readonly root: Graph | Digraph) {}
}
