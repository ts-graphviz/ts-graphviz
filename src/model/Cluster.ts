import {
  ClusterType,
  EdgeTargetLike,
  ICluster,
  IClusterCommonAttributes,
  IContext,
  IEdge,
  INode,
  ISubgraph,
} from '../types';
import { commentOut, concatWordsWithSpace, indent, joinLines } from '../utils/dot-rendering';
import { AttributesBase } from './AttributesBase';
import { ID } from './ID';

/**
 * Base class for clusters.
 * @hidden
 */
export abstract class Cluster<T extends string> extends AttributesBase<T> implements ICluster<T> {
  /** Cluster ID */
  get id(): string | undefined {
    return this.internalID?.value;
  }

  set id(idValue: string | undefined) {
    this.internalID = typeof idValue === 'string' ? new ID(idValue) : undefined;
  }
  /** Comments to include when outputting with toDot. */
  public comment?: string;
  /** The cluster context. */
  public abstract readonly context: IContext;
  /** Indicates the type of cluster. */
  public abstract readonly type: ClusterType;
  /** Common attributes of objects in the cluster. */
  public abstract readonly attributes: Readonly<IClusterCommonAttributes>;

  /**
   * Actual status of ID used when outputting with toDot.
   * @hidden
   */
  private internalID?: ID;

  /**
   * Nodes in the cluster.
   * @hidden
   */
  private nodes: Map<string, INode> = new Map();

  /**
   * Edges in the cluster.
   * @hidden
   */
  private edges: Set<IEdge> = new Set();

  /**
   * Subgraphs in the cluster.
   * @hidden
   */
  private subgraphs: Set<ISubgraph> = new Set();

  /**
   * Add a Node to the cluster.
   */
  public addNode(node: INode): void {
    this.nodes.set(node.id, node);
  }

  /**
   * Add Edge to the cluster.
   */
  public addEdge(edge: IEdge): void {
    this.edges.add(edge);
  }

  /**
   * Add a Subgraph to the cluster.
   */
  public addSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.add(subgraph);
  }

  /**
   * Check if the Node exists in the cluster.
   */
  public existNode(nodeId: string): boolean {
    return this.nodes.has(nodeId);
  }

  /**
   * Check if the Edge exists in the cluster.
   */
  public existEdge(edge: IEdge): boolean {
    return this.edges.has(edge);
  }

  /**
   * Check if the Subgraph exists in the cluster.
   */
  public existSubgraph(subgraph: ISubgraph): boolean {
    return this.subgraphs.has(subgraph);
  }

  /**
   * Create a Subgraph and add it to the cluster.
   */
  public createSubgraph(id?: string): ISubgraph {
    const graph = this.context.createSubgraph(id);
    this.subgraphs.add(graph);
    return graph;
  }

  /**
   * Remove Node from the cluster.
   */
  public removeNode(node: INode | string): void {
    this.nodes.delete(typeof node === 'string' ? node : node.id);
  }

  /**
   * Remove Edge from the cluster.
   */
  public removeEdge(edge: IEdge): void {
    this.edges.delete(edge);
  }

  /**
   * Remove Subgraph from the cluster.
   */
  public removeSubgraph(subgraph: ISubgraph): void {
    this.subgraphs.delete(subgraph);
  }

  /**
   * Create a Node in the cluster.
   */
  public createNode(id: string): INode {
    const node = this.context.createNode(id);
    this.nodes.set(id, node);
    return node;
  }

  /**
   * Get Subgraph in cluster by specifying id.
   *
   * If there is no Subgraph with the specified id in the cluster, return undefined.
   */
  public getSubgraph(id: string): ISubgraph | undefined {
    return Array.from(this.subgraphs.values()).find(subgraph => subgraph.id === id);
  }

  /**
   * Get Node in cluster by specifying id.
   *
   * @description
   * If there is no Node with the specified id in the cluster, return undefined.
   */
  public getNode(id: string): INode | undefined {
    return this.nodes.get(id);
  }

  /** Create Edge and add it to the cluster. */
  public createEdge(target1: EdgeTargetLike, target2: EdgeTargetLike): IEdge;
  public createEdge(...targets: EdgeTargetLike[]): IEdge;
  public createEdge(target1: EdgeTargetLike, target2: EdgeTargetLike, ...targets: EdgeTargetLike[]): IEdge {
    const edge = this.context.createEdge(this, target1, target2, ...targets);
    this.edges.add(edge);
    return edge;
  }

  /**
   * Declarative API for Subgraph.
   *
   * @description
   * If there is a Subgraph with the given ID, use it.
   * If not, create a Subgraph.
   *
   * @param id Subgraph ID.
   * @param callback Callback to operate Subgraph.
   */
  public subgraph(id?: string, callback?: (subgraph: ISubgraph) => void): ISubgraph {
    const subgraph: ISubgraph = id ? this.getSubgraph(id) ?? this.createSubgraph(id) : this.createSubgraph();
    if (callback) {
      callback(subgraph);
    }
    return subgraph;
  }

  /**
   * Declarative API for Node.
   *
   * @description
   * If there is a Node with the given ID, use it.
   * If not, create a Node.
   *
   * @param id Node ID.
   * @param callback Callback to operate Node.
   */
  public node(id: string, callback?: (node: INode) => void): INode {
    const node = this.getNode(id) ?? this.createNode(id);
    if (callback) {
      callback(node);
    }
    return node;
  }

  /**
   * Declarative API for Edge.
   *
   * @param targets Edges.
   * @param callback Callback to operate Edge.
   */
  public edge(targets: EdgeTargetLike[], callback?: (edge: IEdge) => void): IEdge {
    const edge = this.createEdge(...targets);
    if (callback) {
      callback(edge);
    }
    return edge;
  }

  /** Convert Cluster to Dot language. */
  public toDot(): string {
    const comment = this.comment ? commentOut(this.comment) : undefined;
    return joinLines(comment, this.toDotWithoutComment());
  }

  /** @hidden */
  protected toDotWithoutComment(): string {
    const type = this.type;
    const id = this.internalID?.toDot();
    // attributes
    const attributes = Array.from(this.attrs.entries()).map(([key, value]) => `${key} = ${value.toDot()};`);
    const commonAttributes = Object.entries(this.attributes)
      .filter(([_, attrs]) => attrs.size > 0)
      .map(([key, attrs]) => `${key} ${attrs.toDot()};`);

    // objects
    const nodes = Array.from(this.nodes.values()).map(o => o.toDot());
    const subgraphs = Array.from(this.subgraphs.values()).map(o => o.toDot());
    const edges = Array.from(this.edges.values()).map(o => o.toDot());
    const clusterContents = joinLines(...attributes, ...commonAttributes, ...nodes, ...subgraphs, ...edges);
    const dot = joinLines(
      concatWordsWithSpace(type, id, '{'),
      clusterContents ? indent(clusterContents) : undefined,
      '}',
    );
    return dot;
  }
}
