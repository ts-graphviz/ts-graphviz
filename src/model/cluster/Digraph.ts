import { GraphAttributes } from '../attributes/GraphAttributes';
import { Context } from '../Context';
import { RootCluster, RootClusterType } from './Cluster';
/**
 * @category Primary
 */
export class Digraph extends RootCluster<GraphAttributes> {
  public readonly context: Context = new Context(this);
  public type: RootClusterType = 'digraph';
  constructor(id: string = 'G', attributes: GraphAttributes = new GraphAttributes()) {
    super(id, attributes);
  }
}
