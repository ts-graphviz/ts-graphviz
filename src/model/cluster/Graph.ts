import { GraphAttributes } from '../attributes/GraphAttributes';
import { Cluster, GraphType } from './Cluster';
/**
 * @category Primary
 */
export class Graph extends Cluster<GraphAttributes> {
  public type: GraphType = 'graph';
  constructor(id: string = 'G', attributes: GraphAttributes = new GraphAttributes()) {
    super(id, attributes);
  }
  public toDot(): string {
    throw new Error('Method not implemented.');
  }
}
