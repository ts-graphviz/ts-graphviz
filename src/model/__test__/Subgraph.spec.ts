import 'jest-graphviz';
import { DotBase, GraphvizObject } from '../../abstract';
import { IContext, IEdgeTarget, ISubgraph, RootClusterType } from '../../types';
import { AttributesBase } from '../AttributesBase';
import { Cluster } from '../Cluster';
import { Context } from '../Context';
import { Digraph } from '../Digraph';
import { Edge } from '../Edge';
import { Graph } from '../Graph';
import { Node } from '../Node';
import { Subgraph } from '../Subgraph';

const GraphContext = { graphType: RootClusterType.graph } as IContext;

describe('class Subgraph', () => {
  let g: Digraph | Graph;
  let ctx: Context;
  let createEdge: (...targets: IEdgeTarget[]) => Edge;
  const testCases: { title: string; beforeEachFunc: () => void }[] = [
    {
      title: 'root is Digraph',
      beforeEachFunc: () => {
        ctx = new Context();
        g = new Digraph(ctx);
        createEdge = (...targets: IEdgeTarget[]) => new Edge(GraphContext, ...targets);
      },
    },
    {
      title: 'root is Graph',
      beforeEachFunc: () => {
        ctx = new Context();
        g = new Graph(ctx);
        createEdge = (...targets: IEdgeTarget[]) => new Edge(GraphContext, ...targets);
      },
    },
  ];

  testCases.forEach(({ title, beforeEachFunc }) => {
    describe(title, () => {
      beforeEach(beforeEachFunc);
      let subgraph: ISubgraph;

      beforeEach(() => {
        subgraph = g.context.createSubgraph('test');
      });

      it('should be instance of Subgraph/Cluster/AttributesBase/DotBase/GraphvizObject', () => {
        expect(subgraph).toBeInstanceOf(Subgraph);
        expect(subgraph).toBeInstanceOf(Cluster);
        expect(subgraph).toBeInstanceOf(AttributesBase);
        expect(subgraph).toBeInstanceOf(DotBase);
        expect(subgraph).toBeInstanceOf(GraphvizObject);
      });

      test('set attributes', () => {
        subgraph.set('rank', 'same');
        g.addSubgraph(subgraph);
        const dot = g.toDot();
        expect(dot).toBeValidDotAndMatchSnapshot();
      });

      test('set attributes by apply', () => {
        subgraph.apply({
          rank: 'same',
        });
        g.addSubgraph(subgraph);
        const dot = g.toDot();
        expect(dot).toBeValidDotAndMatchSnapshot();
      });

      it('should be subgraph, when subgraph id is "test"', () => {
        subgraph = g.context.createSubgraph('test');
        expect(subgraph.isSubgraphCluster()).toBe(false);
      });

      test.each([
        ['cluster', true],
        ['cluster_hoge', true],
        ['hoge_cluster', false],
        ['example', false],
      ])('if cluster named "%s", isSubgraphCluster should be %p', (id, expected) => {
        subgraph = g.context.createSubgraph(id);
        expect(subgraph.isSubgraphCluster()).toBe(expected);
      });

      describe('subgraph with comment', () => {
        test('single line comment', () => {
          subgraph.comment = 'this is comment.';
          expect(subgraph.toDot()).toMatchSnapshot();
        });

        test('multi line comment', () => {
          subgraph.comment = 'this is comment.\nsecond line.';
          expect(subgraph.toDot()).toMatchSnapshot();
        });
      });

      it('should be subgraph cluster, when subgraph id is "cluster_test"', () => {
        subgraph = g.context.createSubgraph('cluster_test');
        expect(subgraph.isSubgraphCluster()).toBe(true);
      });

      describe('label attribute behavior', () => {
        it('plain text label to be quoted by double quotation', () => {
          subgraph.attributes.graph.set('label', 'this is test for graph label');
          subgraph.attributes.edge.set('label', 'this is test for edge label');
          subgraph.attributes.node.set('label', 'this is test for node label');
          expect(subgraph.toDot()).toMatchSnapshot();
          expect(g.toDot()).toBeValidDot();
        });

        it('html like', () => {
          subgraph.attributes.graph.set('label', '<<B>this is test for graph label</B>>');
          subgraph.attributes.edge.set('label', '<<U>this is test for edge label</U>>');
          subgraph.attributes.node.set('label', '<<I>this is test for node label</I>>');
          expect(subgraph.toDot()).toMatchSnapshot();
          expect(g.toDot()).toBeValidDot();
        });
      });

      describe('addXxx existXxx removeXxx APIs', () => {
        it('Node operation methods works', () => {
          const id = 'node';
          expect(subgraph.existNode(id)).toBe(false);
          const node = new Node(id);
          subgraph.addNode(node);
          expect(subgraph.existNode(id)).toBe(true);
          subgraph.removeNode(node);
          expect(subgraph.existNode(id)).toBe(false);
          subgraph.addNode(node);
          expect(subgraph.existNode(id)).toBe(true);
          subgraph.removeNode(node.id);
          expect(subgraph.existNode(id)).toBe(false);
        });

        it('Edge operation methods works', () => {
          const [node1, node2] = ['node1', 'node2'].map(id => subgraph.createNode(id));
          const edge = createEdge(node1, node2);
          expect(subgraph.existEdge(edge)).toBe(false);
          subgraph.addEdge(edge);
          expect(subgraph.existEdge(edge)).toBe(true);
          expect(subgraph.toDot()).toMatchSnapshot();
          subgraph.removeEdge(edge);
          expect(subgraph.existEdge(edge)).toBe(false);
        });

        it('Subgraph operation methods works', () => {
          const sub = subgraph.context.createSubgraph('sub');
          expect(subgraph.existSubgraph(sub)).toBe(false);
          subgraph.addSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(true);
          expect(subgraph.toDot()).toMatchSnapshot();
          subgraph.removeSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(false);
          subgraph.addSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(true);
          subgraph.removeSubgraph(sub);
          expect(subgraph.existSubgraph(sub)).toBe(false);
        });

        it('should be undefined, when id not set', () => {
          const sub = subgraph.createSubgraph();
          expect(sub.id).toBeUndefined();
          expect(sub.isSubgraphCluster()).toBe(false);
        });

        // test('add/remove/get operation', () => {
        //   const sub = subgraph.context.createSubgraph('sub');
        //   subgraph.add(sub);
        //   expect(subgraph.existSubgraph(sub)).toBe(true);
        //   expect(subgraph.getSubgraph('sub')).toBe(sub);
        //   subgraph.remove(sub);
        //   expect(subgraph.existSubgraph(sub)).toBe(false);

        //   const node = new Node('test');
        //   subgraph.add(node);
        //   expect(subgraph.existNode('test')).toBe(true);
        //   subgraph.remove(node);
        //   expect(subgraph.existNode('test')).toBe(false);

        //   const edge = createEdge(node.port('a'), node.port('b'));
        //   subgraph.add(edge);
        //   expect(subgraph.existEdge(edge)).toBe(true);
        //   subgraph.remove(edge);
        //   expect(subgraph.existEdge(edge)).toBe(false);
        // });

        it('throws an error when the EdgeTarget element is missing', () => {
          const n = subgraph.node('n');
          expect(() => {
            subgraph.edge([]);
          }).toThrow();
          expect(() => {
            subgraph.edge([n]);
          }).toThrow();
        });

        describe('subgraph method', () => {
          it('should be an unnamed Subgraph, when a subgraph is created without specifying an ID.', () => {
            const sub = g.subgraph();
            expect(sub.id).toBeUndefined();
          });

          it('should be same object, when a subgraph with an existing ID is specified', () => {
            const id = 'hoge';
            const sub = g.createSubgraph(id);
            expect(g.subgraph(id)).toBe(sub);
          });
        });
      });
    });
  });
});
