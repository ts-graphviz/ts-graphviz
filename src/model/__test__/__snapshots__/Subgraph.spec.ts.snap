// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Subgraph root is Digraph addXxx existXxx removeXxx APIs Edge operation methods works 1`] = `
"subgraph test {
  node1;
  node2;
  node1 -- node2;
}"
`;

exports[`class Subgraph root is Digraph addXxx existXxx removeXxx APIs Subgraph operation methods works 1`] = `
"subgraph test {
  subgraph sub {
  }
}"
`;

exports[`class Subgraph root is Digraph label attribute behavior html like 1`] = `
"subgraph test {
  graph [
    label = <<B>this is test for graph label</B>>,
  ];
  edge [
    label = <<U>this is test for edge label</U>>,
  ];
  node [
    label = <<I>this is test for node label</I>>,
  ];
}"
`;

exports[`class Subgraph root is Digraph label attribute behavior plain text label to be quoted by double quotation 1`] = `
"subgraph test {
  graph [
    label = \\"this is test for graph label\\",
  ];
  edge [
    label = \\"this is test for edge label\\",
  ];
  node [
    label = \\"this is test for node label\\",
  ];
}"
`;

exports[`class Subgraph root is Digraph set attributes 1`] = `
"digraph {
  subgraph test {
    rank = same;
  }
}"
`;

exports[`class Subgraph root is Digraph set attributes by apply 1`] = `
"digraph {
  subgraph test {
    rank = same;
  }
}"
`;

exports[`class Subgraph root is Digraph subgraph with comment multi line comment 1`] = `
"// this is comment.
// second line.
subgraph test {
}"
`;

exports[`class Subgraph root is Digraph subgraph with comment single line comment 1`] = `
"// this is comment.
subgraph test {
}"
`;

exports[`class Subgraph root is Graph addXxx existXxx removeXxx APIs Edge operation methods works 1`] = `
"subgraph test {
  node1;
  node2;
  node1 -- node2;
}"
`;

exports[`class Subgraph root is Graph addXxx existXxx removeXxx APIs Subgraph operation methods works 1`] = `
"subgraph test {
  subgraph sub {
  }
}"
`;

exports[`class Subgraph root is Graph label attribute behavior html like 1`] = `
"subgraph test {
  graph [
    label = <<B>this is test for graph label</B>>,
  ];
  edge [
    label = <<U>this is test for edge label</U>>,
  ];
  node [
    label = <<I>this is test for node label</I>>,
  ];
}"
`;

exports[`class Subgraph root is Graph label attribute behavior plain text label to be quoted by double quotation 1`] = `
"subgraph test {
  graph [
    label = \\"this is test for graph label\\",
  ];
  edge [
    label = \\"this is test for edge label\\",
  ];
  node [
    label = \\"this is test for node label\\",
  ];
}"
`;

exports[`class Subgraph root is Graph set attributes 1`] = `
"graph {
  subgraph test {
    rank = same;
  }
}"
`;

exports[`class Subgraph root is Graph set attributes by apply 1`] = `
"graph {
  subgraph test {
    rank = same;
  }
}"
`;

exports[`class Subgraph root is Graph subgraph with comment multi line comment 1`] = `
"// this is comment.
// second line.
subgraph test {
}"
`;

exports[`class Subgraph root is Graph subgraph with comment single line comment 1`] = `
"// this is comment.
subgraph test {
}"
`;
